File: sdk-typescript/packages/slides/src/slides.ts
/**
 * WAVE SDK - Slides API
 *
 * Presentation to video conversion with narration and transitions.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export type ConversionStatus = "pending" | "processing" | "ready" | "failed";
export type SlideFormat = "pptx" | "pdf" | "google_slides" | "keynote";
export type TransitionPreset = "none" | "fade" | "slide" | "zoom" | "morph";

export interface Conversion extends Timestamps {
  id: string;
  organization_id: string;
  title: string;
  status: ConversionStatus;
  input_format: SlideFormat;
  input_url: string;
  output_url?: string;
  slide_count: number;
  duration_seconds?: number;
  resolution: string;
  narration_enabled: boolean;
  progress_percent: number;
  error?: string;
}
export interface SlideNarration {
  slide_index: number;
  text: string;
  voice_id?: string;
  duration_seconds?: number;
}
export interface ConvertRequest {
  title: string;
  input_url: string;
  input_format: SlideFormat;
  resolution?: string;
  narration?: SlideNarration[];
  transition?: TransitionPreset;
  slide_duration_seconds?: number;
  webhook_url?: string;
}

/**
 * Presentation-to-video conversion with narration and transitions.
 *
 * @example
 * ```typescript
 * const conversion = await wave.slides.convert({ title: "Q4 Review", input_url: "https://...", input_format: "pptx" });
 * const ready = await wave.slides.waitForReady(conversion.id);
 * ```
 */
export class SlidesAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/slides";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async convert(request: ConvertRequest): Promise<Conversion> {
    return this.client.post<Conversion>(this.basePath, request);
  }
  async get(conversionId: string): Promise<Conversion> {
    return this.client.get<Conversion>(`${this.basePath}/${conversionId}`);
  }
  async list(params?: PaginationParams): Promise<PaginatedResponse<Conversion>> {
    return this.client.get<PaginatedResponse<Conversion>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async remove(conversionId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${conversionId}`);
  }
  async getProgress(
    conversionId: string,
  ): Promise<{ status: ConversionStatus; progress_percent: number }> {
    return this.client.get<{ status: ConversionStatus; progress_percent: number }>(
      `${this.basePath}/${conversionId}/progress`,
    );
  }
  async addNarration(conversionId: string, narrations: SlideNarration[]): Promise<Conversion> {
    return this.client.post<Conversion>(`${this.basePath}/${conversionId}/narration`, {
      narrations,
    });
  }

  async waitForReady(
    conversionId: string,
    options?: { pollInterval?: number; timeout?: number },
  ): Promise<Conversion> {
    const pollInterval = options?.pollInterval || 3000;
    const timeout = options?.timeout || 600000;
    const startTime = Date.now();
    while (Date.now() - startTime < timeout) {
      const conversion = await this.get(conversionId);
      if (conversion.status === "ready") return conversion;
      if (conversion.status === "failed")
        throw new Error(`Conversion failed: ${conversion.error || "Unknown"}`);
      await new Promise((r) => setTimeout(r, pollInterval));
    }
    throw new Error(`Conversion timed out after ${timeout}ms`);
  }
}

export function createSlidesAPI(client: WaveClient): SlidesAPI {
  return new SlidesAPI(client);
}

File: sdk-python/wave/slides.py
"""WAVE SDK - Slides API. Presentation to video conversion."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Conversion(BaseModel):
    id: str; organization_id: str; title: str; status: str; input_format: str; input_url: str; output_url: str | None = None; slide_count: int = 0; duration_seconds: float | None = None; progress_percent: int = 0; error: str | None = None; created_at: str; updated_at: str

class SlidesAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/slides"
    def convert(self, title: str, input_url: str, input_format: str, **kwargs: Any) -> Conversion: return Conversion(**self._client.post(self._base, json={"title": title, "input_url": input_url, "input_format": input_format, **kwargs}))
    def get(self, conversion_id: str) -> Conversion: return Conversion(**self._client.get(f"{self._base}/{conversion_id}"))
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def remove(self, conversion_id: str) -> None: self._client.delete(f"{self._base}/{conversion_id}")
    def add_narration(self, conversion_id: str, narrations: list[dict]) -> Conversion: return Conversion(**self._client.post(f"{self._base}/{conversion_id}/narration", json={"narrations": narrations}))
    def wait_for_ready(self, conversion_id: str, poll_interval: float = 3.0, timeout: float = 600.0) -> Conversion:
        start = time.time()
        while time.time() - start < timeout:
            c = self.get(conversion_id)
            if c.status == "ready": return c
            if c.status == "failed": raise RuntimeError(f"Conversion failed: {c.error}")
            time.sleep(poll_interval)
        raise TimeoutError(f"Conversion timed out after {timeout}s")
